ssl/statem/statem_lib.c

```diff
@@ -1633,6 +1633,55 @@ int tls_get_message_header(SSL_CONNECTION *s, int *mt)
     return 1;
 }
 
+int tls_common_finish_mac(SSL_CONNECTION *s) {
+    unsigned char *msg = (unsigned char *)s->init_buf->data;
+    size_t msg_len = s->init_num;
+    size_t hdr_len = 0;
+
+    if (!SSL_CONNECTION_IS_DTLS(s) && !RECORD_LAYER_is_sslv2_record(&s->rlayer))
+        hdr_len = SSL3_HM_HEADER_LENGTH;
+
+    if (SSL_CONNECTION_IS_DTLS(s) && s->version != DTLS1_BAD_VER)
+        hdr_len = DTLS1_HM_HEADER_LENGTH;
+
+    if (SSL_CONNECTION_IS_DTLS(s) && s->version == DTLS1_BAD_VER)
+        msg += DTLS1_HM_HEADER_LENGTH;
```

Comment:
I would instead do:
```C
    if (SSL_CONNECTION_IS_DTLS(s)) {
        if (s->version != DTLS1_BAD_VER)
            hdr_len = DTLS1_HM_HEADER_LENGTH;
        else
            msg += DTLS1_HM_HEADER_LENGTH;
    } else if (!RECORD_LAYER_is_sslv2_record(&s->rlayer)) {
        hdr_len = SSL3_HM_HEADER_LENGTH;
    }
```

Reply:
Sorry I missed this comment. I like your approach. Fixed now.